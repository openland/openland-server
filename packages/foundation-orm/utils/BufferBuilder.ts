export class BufferBuilder {
    
    storage: Buffer;
    used: number = 0;

    constructor(capacity: number = 64) {
        this.storage = Buffer.alloc(capacity);
    }

    make() {
        const result = Buffer.alloc(this.used);
        this.storage.copy(result, 0, 0, this.used);
        return result;
    }

    need(numBytes: number) {
        if (this.storage.length < this.used + numBytes) {
            let newAmt = this.storage.length;
            while (newAmt < this.used + numBytes) {
                newAmt *= 2;
            }
            const newStorage = Buffer.alloc(newAmt);
            this.storage.copy(newStorage);
            this.storage = newStorage;
        }
    }

    appendByte(val: number) {
        this.need(1); this.storage[this.used++] = val;
    }

    appendString(val: string) {
        const len = Buffer.byteLength(val);
        this.need(len);
        this.storage.write(val, this.used);
        this.used += len;
    }

    appendBuffer(val: Buffer) {
        this.need(val.length);
        val.copy(this.storage, this.used);
        this.used += val.length;
    }
}