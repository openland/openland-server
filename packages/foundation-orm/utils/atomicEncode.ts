export function encodeAtomic(value: number) {
    const b = Buffer.alloc(4);
    b.writeInt32LE(value, 0);
    return b;
}

export function decodeAtomic(src: Buffer): number {
    return src.readInt32LE(0);
}