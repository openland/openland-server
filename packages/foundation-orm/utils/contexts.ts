import { createContextNamespace, Context } from 'openland-utils/Context';
import { FTransaction } from 'foundation-orm/FTransaction';
import { FConnection } from 'foundation-orm/FConnection';
import { FCacheContext } from 'foundation-orm/FCacheContext';

export const FTransactionContext = createContextNamespace<FTransaction | null>('tx', null);
export const FCacheContextContext = createContextNamespace<FCacheContext | null>('tx-cache', null);

export function resolveContext(ctx: Context) {
    let tx = FTransactionContext.get(ctx);
    if (tx) {
        return tx;
    }
    // let cache = FCacheContextContext.get(ctx);
    // if (cache) {
    //     return cache;
    // }
    return FConnection.globalContext;
}