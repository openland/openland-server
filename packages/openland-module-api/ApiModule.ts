import { ApolloClient } from 'apollo-client';
import { InMemoryCache } from 'apollo-cache-inmemory';
import { SchemaLink } from 'apollo-link-schema';
import { serverRoleEnabled } from 'openland-utils/serverRoleEnabled';
import { initApi } from './initApi';
import { initHealthcheck } from './initHealthcheck';
import { injectable } from 'inversify';
import { GQL_SPEC_VERSION } from './schema/SchemaSpec';
import { getSchemeVersion } from './schema/SchemaSpecGenerator';
import { buildSchema } from '../openland-graphql/buildSchema';
import { Schema } from './schema/Schema';
import { AuthContext } from 'openland-module-auth/AuthContext';
import { CacheContext } from './CacheContext';
import { FDB, Store } from 'openland-module-db/FDB';
import { AppContext } from 'openland-modules/AppContext';
import { createNamedContext } from '@openland/context';
import { withLogMeta, createLogger } from '@openland/log';

const defaultCtx = createNamedContext('ctx');
const logger = createLogger('api-module');

@injectable()
export class ApiModule {
    
    start = async () => {
        logger.log(defaultCtx, 'Start API Module');
        if (serverRoleEnabled('api')) {
            let currentSchemeSpecVersion = getSchemeVersion(buildSchema(__dirname + '/../'));

            if (GQL_SPEC_VERSION !== currentSchemeSpecVersion) {
                throw new Error(`Schema version mismatch expected ${GQL_SPEC_VERSION} got ${currentSchemeSpecVersion}, did you forgot to run yarn schema:gen ?`);
            }
            await initApi(false);
        } else {
            if (!serverRoleEnabled('admin')) {
                await initHealthcheck();
            }
        }
    }

    createClientForUser = async (email: string) => {
        let uid = (await FDB.User.findAll(defaultCtx)).find((v) => v.email === email)!.id;
        let tid = (await Store.AuthToken.findAll(defaultCtx)).find((v) => v.uid === uid)!.uuid;
        return this.createClient({ uid, tid });
    }

    createClient = (args: { uid?: number, tid?: string }) => {
        let ctx = defaultCtx;
        if (args.uid && args.tid) {
            ctx = AuthContext.set(ctx, { uid: args.uid, tid: args.tid });
            ctx = withLogMeta(ctx, { uid: args.uid, tid: args.tid });
        }
        ctx = CacheContext.set(ctx, new Map());
        return new ApolloClient({
            cache: new InMemoryCache(),
            link: new SchemaLink({ schema: Schema(process.env.TESTING === 'true'), context: new AppContext(ctx) })
        });
    }
}