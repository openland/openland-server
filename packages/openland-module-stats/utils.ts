export const resizeUcarecdnImage = (imageStr: string, newSize: { width: number; height: number }) => {
    // ["https:", "", "ucarecdn.com", "25629a3c-1ebe-4d49-8560-9df3b92ade3a", "-", "resize", "80x80", ""]
    const [, , , ucareImageID] = imageStr.split('/');
    return `ttps://ucarecdn.com/${ucareImageID}/-/resize/${newSize.width}x${newSize.height}/`;
};
