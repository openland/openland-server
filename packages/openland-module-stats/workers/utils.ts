import { Context } from '@openland/context';
import { Modules } from '../../openland-modules/Modules';
import { Store } from '../../openland-module-db/FDB';
import { MessageInput } from '../../openland-module-messaging/MessageInput';
import { boldString, buildMessage, heading, insaneString } from '../../openland-utils/MessageBuilder';
import { inTx } from '@openland/foundationdb';
import moment from 'moment';

export const getSuperNotificationsBotId = (ctx: Context) => Modules.Super.getEnvVar<number>(ctx, 'super-notifications-app-id');
export const getOnboardingReportsChatId = (ctx: Context) => Modules.Super.getEnvVar<number>(ctx, 'onboarding-reports-chat-id');
export const getEngagementReportsChatId = (ctx: Context) => Modules.Super.getEnvVar<number>(ctx, 'engagement-reports-chat-id');
export const getLeaderboardsChatId = (ctx: Context) => Modules.Super.getEnvVar<number>(ctx, 'leaderboards-chat-id');

export const getOnboardingCounters = async (ctx: Context, startDate: number) => {
    let activationsData = await Modules.Search.search(ctx, {
        index: 'hyperlog', type: 'hyperlog', // scroll: '1m',
        body: {
            query: {
                bool: {
                    must: [{ term: { type: 'user_activated' } }, { term: { ['body.isTest']: false } }, {
                        range: {
                            date: {
                                gte: startDate,
                            },
                        },
                    }],
                },
            },
        }, size: 0,
    });
    let newUserEntrances = (activationsData.hits.total as any).value;

    const newMobileUsersQuery = await Modules.Search.search(ctx, {
        index: 'hyperlog', type: 'hyperlog', body: {
            query: {
                bool: {
                    must: [{ term: { type: 'new-mobile-user' } }, { term: { ['body.isTest']: false } }, {
                        range: {
                            date: {
                                gte: startDate,
                            },
                        },
                    }],
                },
            },
        }, size: 0,
    });
    const newMobileUsers = (newMobileUsersQuery.hits.total as any).value;

    const newSendersQuery = await Modules.Search.search(ctx, {
        index: 'hyperlog', type: 'hyperlog', body: {
            query: {
                bool: {
                    must: [{ term: { type: 'new-sender' } }, { term: { ['body.isTest']: false } }, {
                        range: {
                            date: {
                                gte: startDate,
                            },
                        },
                    }],
                },
            },
        }, size: 0,
    });
    const newSenders = (newSendersQuery.hits.total as any).value;

    const newInvitersQuery = await Modules.Search.search(ctx, {
        index: 'hyperlog', type: 'hyperlog', body: {
            query: {
                bool: {
                    must: [{ term: { type: 'new-inviter' } }, { term: { ['body.isTest']: false } }, {
                        range: {
                            date: {
                                gte: startDate,
                            },
                        },
                    }],
                },
            },
        }, size: 0,
    });
    const newInviters = (newInvitersQuery.hits.total as any).value;

    const newThreeLikeGiversQuery = await Modules.Search.search(ctx, {
        index: 'hyperlog', type: 'hyperlog', body: {
            query: {
                bool: {
                    must: [{ term: { type: 'new-three-like-giver' } }, { term: { ['body.isTest']: false } }, {
                        range: {
                            date: {
                                gte: startDate,
                            },
                        },
                    }],
                },
            },
        }, size: 0,
    });
    const newThreeLikeGivers = (newThreeLikeGiversQuery.hits.total as any).value;

    const newThreeLikeGettersQuery = await Modules.Search.search(ctx, {
        index: 'hyperlog', type: 'hyperlog', body: {
            query: {
                bool: {
                    must: [{ term: { type: 'new-three-like-getter' } }, { term: { ['body.isTest']: false } }, {
                        range: {
                            date: {
                                gte: startDate,
                            },
                        },
                    }],
                },
            },
        }, size: 0,
    });
    const newThreeLikeGetters = (newThreeLikeGettersQuery.hits.total as any).value;

    let newAboutFillersData = await Modules.Search.search(ctx, {
        index: 'hyperlog', type: 'hyperlog', // scroll: '1m',
        body: {
            query: {
                bool: {
                    must: [{ term: { type: 'new-about-filler' } }, { term: { ['body.isTest']: false } }, {
                        range: {
                            date: {
                                gte: startDate,
                            },
                        },
                    }],
                },
            }, aggs: {
                usersCount: {
                    cardinality: {
                        field: 'body.uid',
                    },
                },
            },
        }, size: 0,
    });
    let newAboutFillers = newAboutFillersData.aggregations.usersCount.value;

    return {
        newUserEntrances,
        newMobileUsers,
        newSenders,
        newInviters,
        newAboutFillers,
        newThreeLikeGivers,
        newThreeLikeGetters,
    };
};

export const getEngagementCounters = async (ctx: Context, startDate: number) => {
    let activesData = await Modules.Search.search(ctx, {
        index: 'hyperlog', type: 'hyperlog', // scroll: '1m',
        body: {
            query: {
                bool: {
                    must: [{ term: { type: 'presence' } }, { term: { ['body.online']: true } }, {
                        range: {
                            date: {
                                gte: startDate,
                            },
                        },
                    }],
                },
            }, aggs: {
                actives: {
                    cardinality: {
                        field: 'body.uid',
                    },
                },
            },
        }, size: 0,
    });

    let actives = activesData.aggregations.actives.value;

    let sendersData = await Modules.Search.search(ctx, {
        index: 'message', type: 'message',
        body: {
            query: {
                bool: {
                    must: [{ term: { isService: false } }, {
                        range: {
                            createdAt: {
                                gte: startDate,
                            },
                        },
                    }],
                },
            }, aggs: {
                senders: {
                    cardinality: {
                        field: 'uid',
                    },
                },
            },
        }, size: 0,
    });

    let senders = sendersData.aggregations.senders.value;
    let messagesSent = (sendersData.hits.total as any).value;

    let todayLikersData = await Modules.Search.search(ctx, {
        index: 'hyperlog', type: 'hyperlog', // scroll: '1m',
        body: {
            query: {
                bool: {
                    must: [{ term: { type: 'new-reaction' } }, {
                        range: {
                            date: {
                                gte: startDate,
                            },
                        },
                    }, { term: { ['body.isTest']: false } }],
                },
            }, aggs: {
                givers: {
                    cardinality: {
                        field: 'body.uid',
                    },
                },
                getters: {
                    cardinality: {
                        field: 'body.messageAuthorId',
                    },
                },
            },
        }, size: 0,
    });
    let todayLikeGivers = todayLikersData.aggregations.givers.value;
    let todayLikeGetters = todayLikersData.aggregations.getters.value;

    let callsData = await Modules.Search.search(ctx, {
        index: 'hyperlog', type: 'hyperlog',
        body: {
            query: {
                bool: {
                    must: [{ term: { type: 'call_ended' } }, {
                        range: {
                            date: {
                                gte: startDate,
                            },
                        },
                    }],
                },
            }, aggs: {
                totalDuration: {
                    sum: {
                        field: 'body.duration',
                    },
                },
            },
        }, size: 0,
    });
    let totalCallsDuration = callsData.aggregations.totalDuration.value;

    return {
        actives,
        senders,
        todayLikeGivers,
        todayLikeGetters,
        messagesSent,
        totalCallsDuration: Math.round(moment.duration(totalCallsDuration).asMinutes())
    };
};

export const alertIfRecord = async (
    parent: Context,
    cid: number,
    metricName: string,
    value: number,
    build: (prevRecord: number, currentRecord: number) => MessageInput,
) => {
    return await inTx(parent, async ctx => {
        let botId = await getSuperNotificationsBotId(ctx);
        if (!botId) {
            return;
        }

        let metric = await Store.StatsRecords.byId(metricName);
        let prevValue = await metric.get(ctx);
        if (value > prevValue) {
            metric.set(ctx, value);

            await Modules.Messaging.sendMessage(ctx, cid, botId, {
                ...build(prevValue, value),
            });
        }
    });
};

export const alertIfRecordDelta = async (
    parent: Context,
    cid: number,
    metricName: string,
    value: number,
    build: (prevRecord: number, currentRecord: number) => MessageInput,
) => {
    return await inTx(parent, async ctx => {
        let botId = await getSuperNotificationsBotId(ctx);
        if (!botId) {
            return;
        }

        let metric = await Store.StatsRecords.byId(metricName);
        let metricDelta = await Store.StatsRecords.byId(metricName + '-delta');

        let prevValue = await metric.get(ctx);
        let maxDelta = await metricDelta.get(ctx);

        metric.set(ctx, value);
        if ((value - prevValue) > maxDelta) {
            metricDelta.set(ctx, value - prevValue);

            await Modules.Messaging.sendMessage(ctx, cid, botId, {
                ...build(maxDelta, value - prevValue),
            });
        }
    });
};

export const buildDailyRecordAlert = (metricTitle: string) => {
    return (prevRecord: number, currentRecord: number) => buildMessage(heading(`${metricTitle} count have reached `,
        insaneString(currentRecord.toString()), ' today!',
    ), '\nPrevious record was ', boldString(prevRecord.toString()));
};

export const buildWeeklyRecordAlert = (metricTitle: string) => {
    return (prevRecord: number, currentRecord: number) => buildMessage(heading(`${metricTitle} count have reached `,
        insaneString(currentRecord.toString()), ' this week!',
    ), '\nPrevious record was ', boldString(prevRecord.toString()));
};