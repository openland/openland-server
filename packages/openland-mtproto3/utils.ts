import { DocumentNode, getOperationAST } from 'graphql';

export function isAsyncIterator(iterator: any): iterator is AsyncIterable<any> {
    return iterator.hasOwnProperty(Symbol.asyncIterator);
}

export function isSubscriptionQuery(query: DocumentNode, operationName?: string) {
    const operationAST = getOperationAST(query, operationName);

    return !!operationAST && operationAST.operation === 'subscription';
}