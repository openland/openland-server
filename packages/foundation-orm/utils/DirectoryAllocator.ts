import { FConnection } from 'foundation-orm/FConnection';
import { FKeyEncoding } from './FKeyEncoding';
import { createLogger } from 'openland-log/createLogger';
import { backoff } from 'openland-server/utils/timer';

const rootPrefix = Buffer.from('f0', 'hex');
const dataPrefix = Buffer.concat([rootPrefix, Buffer.from('01', 'hex')]);
const metaPrefix = Buffer.concat([rootPrefix, Buffer.from('fe', 'hex')]);
const regsPrefix = Buffer.concat([metaPrefix, Buffer.from('01', 'hex')]);
const counterKey = Buffer.concat([metaPrefix, Buffer.from('02', 'hex')]);
const logger = createLogger('directory-allocator');

function buildDataPrefix(counter: number) {
    let res = new Buffer(2);
    res.writeInt16BE(counter, 0);
    return Buffer.concat([dataPrefix, res]);
}

export class DirectoryAllocator {
    readonly connection: FConnection;

    constructor(connection: FConnection) {
        this.connection = connection;
    }

    async allocateDirectory(key: (string | number | boolean)[]) {
        logger.debug('Allocating for key: ' + JSON.stringify(key));
        let destKey = Buffer.concat([regsPrefix, FKeyEncoding.encodeKey([...key])]);
        try {
            return await backoff(async () => await this.connection.fdb.doTransaction(async (tx) => {
                let res = await tx.get(destKey);
                if (res) {
                    return buildDataPrefix(res.value as number);
                }
                let nextPrefix = await tx.get(counterKey);
                let nextCounter = 1;
                if (!nextPrefix) {
                    tx.set(counterKey, { value: 1 });
                } else {
                    nextCounter = (nextPrefix.value as number) + 1;
                    tx.set(counterKey, { value: nextPrefix + 1 });
                }
                if (nextCounter > 6536) {
                    throw Error('Key space overflowed');
                }
                tx.set(destKey, { value: nextCounter });
                logger.debug('Allocated ' + nextCounter + ' for key: ' + JSON.stringify(key));
                return buildDataPrefix(nextCounter);
            }));
        } catch (e) {
            console.warn(e);
            throw Error('Unable to allocate key!');
        }
    }
}