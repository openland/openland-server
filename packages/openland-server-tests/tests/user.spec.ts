import { prepareServer } from './utils/prepareServer';
import { ApiModule } from 'openland-module-api/ApiModule';
import gql from 'graphql-tag';

describe('User', () => {
    let api: ApiModule;
    beforeAll(async () => {
        api = await prepareServer();
    });
    it('should return own profile', async () => {
        let client = await api.createClientForUser('test1111@openland.com');
        let res = (await client.query({
            query: gql`
            query Me {
                me {
                    id
                    isBot
                    isYou
                    firstName
                    lastName
                    about
                    phone
                    email
                    website
                    linkedin
                    twitter
                    location
                    shortname
                }
            }
        ` })).data as any;
        expect(res.me).toMatchSnapshot();
    });
    it('should return null for non-logged in user', async () => {
        let client = await api.createClient({});
        let res = (await client.query({
            query: gql`
            query Me {
                me {
                    id
                    isBot
                    isYou
                    firstName
                    lastName
                    about
                    phone
                    email
                    website
                    linkedin
                    twitter
                    location
                    shortname
                }
            }
        ` })).data as any;
        expect(res.me).toBeNull();
    });
});