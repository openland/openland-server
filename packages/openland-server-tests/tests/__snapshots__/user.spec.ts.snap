// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User should return own profile 1`] = `
Object {
  "__typename": "User",
  "about": "Central Marketing Coordinator",
  "email": "test1111@openland.com",
  "firstName": "Maurine",
  "id": "GvNrAkJ7rbFGQmJJJJdyhKaBr4",
  "lastName": "Rath",
  "phone": "(426) 577-4479",
}
`;
